### ISSUE TITLE
File inclusion classification unclear

### ISSUE THREAD

**Issue by ordinaehs** [OWASP/ASVS#112]
Both V5.13 and V16.4 mention File Inclusion. The way I read it, V5.13 is only when the data sent is a path to an already existing resource (either local or remote), and V16.4 is when the data sent is code or a file. Assuming I got this correct, can this difference please be made more clear in the descriptions?

**Comment by vanderaj** [comment]
Let's have a discussion of how best to address this:

V5 is about input validation and output encoding as a first line of defence to prevent the issue in the first place, even if configuration (V16) is weak
V16 is about improving or altering configuration to protect the platform, even if the application's first line of defence (V5) is weak

We try to de-dupe as much as possible, which in ASVS 3.0, meant a lot of retired duplicate requirements. 

If both requirements in 3.0.1 are asking for the same resolution for file inclusion, then we need to eliminate one of them. 

If they are two controls for different activities, then how best do we change the text to make it clear?

Do we need to make LFI and RFI seperate in V16? I believe that V5.13 could be changed to encompass both the metadata of the file (path, filename, etc) as well as the contents without duplication

Thoughts?

**Comment by vanderaj** [comment]
Additionally, the ASVS doesn't cope with this "new" reflective file download attack:

https://www.trustwave.com/Resources/SpiderLabs-Blog/Reflected-File-Download---A-New-Web-Attack-Vector/

Do we need to cope with this? I must admit to not seeing this in play before, but Trustwave's paper is fairly clear on how to exploit it.

**Comment by ordinaehs** [comment]
Might I suggest changing the wording of V16.4 to:
Verify that untrusted data is not used within inclusion, class loader, or reflection capabilities to prevent remote/local code execution vulnerabilities.

This means that local/remote file inclusion is only part of V5.13, while V16.4 is more about getting the application to execute untrusted code, or trusted code in an unwanted way.

That reflective file download should probably be considered a case of V5.12?

**Comment by vanderaj** [comment]
Done. I've updated the text as you described for 16.4. I've also included 16.2 to include reflective file download, as that's specifically about not trusting file metadata supplied by the user. I'm trying to keep section 16 about files and resources, and stop polluting V5 with everything under the sun. 

16.2	Verify that untrusted file data submitted to the application is not used directly with file I/O commands, particularly to protect against path traversal, local file include, file mime type, reflective file download, and OS command injection vulnerabilities.

16.4	Verify that untrusted data is not used within inclusion, class loader, or reflection capabilities to prevent remote/local code execution vulnerabilities.
